### DISCUSSION TITLE
take off issue

### DISCUSSION THREAD

**Discussion by WangMahua** [bitcraze/discussions#934] [Q&A]
Hello everyone,

I'm new to Crazyflie. I've been trying to control Crazyflie 2.0 with ROS1 on Ubuntu 20.04 using Python 3. I have three PAs connected to three Crazyflies, respectively. I'm using OptiTrack to detect the position of the Crazyflies.

I referred to this [repository](https://github.com/JGSuw/rospy_crazyflie) for controlling the Crazyflies. I'm aware that the repository is designed for Python 2, so I've been attempting to fix some errors to make it work with Python 3.

I can successfully connect to each Crazyflie, but when I run my code and send some commands, it doesn't take off properly every time. Typically, after I run my code, the Crazyflie's LED goes out and then reconnects. Afterward, the red LED stays on. Sometimes, when I call the function "_set_vel_setpoint()", the propellers rotate but the Crazyflie doesn't move to the correct position, even though the command I send is correct. I'm pretty sure the propeller directions are set correctly.

Here's my code:
```python =
import rospy
import numpy as np
import rospy_crazyflie.crazyflie_client as crazyflie_client
from geometry_msgs.msg import Twist,Pose,TwistStamped,PoseStamped
from pyquaternion import Quaternion
from std_msgs.msg import Float64MultiArray,String
from gazebo_msgs.msg import ModelStates

class cf_controller:
	def __init__(self, uavtype, idx):
		self.name = uavtype
		self.idx = idx
		self.theta = None

		crazyflies = crazyflie_client.get_crazyflies(server='/crazyflie_server')
		self.client_name = crazyflie_client.CrazyflieClient(crazyflies[idx])

       # I'm only including the relevant code related to the issue.
       # cf_ini_state represents the target position, and cf_state represents the current position. Both of them provide correct information.

	def takeoff(self):
		self.client_name.take_off(.3)

	def hover(self):
		self.client_name._set_vel_setpoint(0.5*(self.cf_ini_state[0]-self.cf_state[0]),0.5*(self.cf_ini_state[1]-self.cf_state[1]),0.5*(self.cf_ini_state[2]-self.cf_state[2]),-0.5*(self.cf_ini_state[3]-self.cf_state[3])*180/np.pi)

	def process(self):
		self.client_name._set_vel_setpoint(self.cmd[0],self.cmd[1],self.cmd[2],self.cmd[3].X*180/np.pi)


if __name__ == "__main__":
	rospy.init_node('cfcontroller')
	rate = rospy.Rate(100)

	uavtype_list = np.array(rospy.get_param("/agent/name")) 
	target_list = np.array(rospy.get_param("/env/target_list")) 

	crazyflies = crazyflie_client.get_crazyflies(server='/crazyflie_server')
	cf_list = []

	for i in range(len(uavtype_list)):
		cf_list.append(cf_controller(uavtype_list[i],i))

	# for i in range(len(uavtype_list)):
	for i in range(3):
		cf = cf_list[i] 
		cf.takeoff()

	print("wait takeoff")
	while rospy.get_param(uavtype_list[0] + "_is_flying") == 0 or rospy.get_param(uavtype_list[1] + "_is_flying") == 0 or rospy.get_param(uavtype_list[2] + "_is_flying") == 0:
		rate = rospy.Rate(100)

	print("start to hover")
	while rospy.get_param("start_control") == 0:
		cf = cf_list[i] 
		cf.hover()

```

Do you have any ideas on how I could solve this issue?

**Comment by knmcguire** [comment]
Hi! 

Hmmm we are not aware of this repository, nor can we really see what is going on, since the functionality is a bit hidden in the python library that somebody else has written.

Could you try out the regular python library instead? Like the motion commander autonomy demo? https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/motion_commander_demo.py
